lib/SIL/SILBasicBlock.cpp

```diff
@@ -187,9 +187,32 @@ transferNodesFromList(llvm::ilist_traits<SILBasicBlock> &SrcTraits,
   // If splicing blocks not in the same function, update the parent pointers.
   for (; First != Last; ++First) {
     First->Parent = Parent;
-    for (auto &II : *First)
+    for (auto &II : *First) {
       II.setDebugScope(B,
                        ScopeCloner.getOrCreateClonedScope(II.getDebugScope()));
+      SILInstruction *I = &II;
+      // Maintain the registrations of opened archetypes.
+      if (isa<OpenExistentialAddrInst>(I) || isa<OpenExistentialRefInst>(I) ||
+          isa<OpenExistentialBoxInst>(I)) {
+        auto Ty = I->getType().getSwiftRValueType();
+        if (isa<ArchetypeType>(Ty) &&
+            !cast<ArchetypeType>(Ty)->getOpenedExistentialType().isNull()) {
+          SrcTraits.Parent->removeOpenedArchetypeDef(Ty, I);
+          Parent->addOpenedArchetypeDef(Ty, I);
+        }
+      } else if (isa<OpenExistentialMetatypeInst>(I)) {
+        SILType InstanceTy = I->getType();
+        while (isa<AnyMetatypeType>(InstanceTy.getSwiftRValueType())) {
+          InstanceTy = InstanceTy.getMetatypeInstanceType(I->getModule());
+        }
+        auto Ty = InstanceTy.getSwiftRValueType();
+        if (isa<ArchetypeType>(Ty) &&
+            !cast<ArchetypeType>(Ty)->getOpenedExistentialType().isNull()) {
+          SrcTraits.Parent->removeOpenedArchetypeDef(Ty, I);
+          Parent->addOpenedArchetypeDef(Ty, I);
+        }
+      }
+    }
```

Comment:
In this hypothetical world of type def-use, type-defs would be additional function arguments and values produced by open_existential. type-use would be operands on generic witness_method, apply, try_apply, and some others. It would be nice to have a full list of instructions that consume archetypes. I'll guess that it includes these (quickly scanning SIL.rst):
- alloc_stack/box
- *_method
- partial_apply
- *metatype
- checked_cast*
- unchecked_ref_cast* maybe
- unchecked_addr_cast maybe

Obviously we would measure the memory increase caused by the extra arguments and operands before designing something like this.